lms/djangoapps/lti_provider/outcomes.py

```diff
@@ -0,0 +1,199 @@
+    sourced_id = etree.SubElement(sourced_guid, 'sourcedId')
+    result = etree.SubElement(result_record, 'result')
+    result_score = etree.SubElement(result, 'resultScore')
+    language = etree.SubElement(result_score, 'language')
+    text_string = etree.SubElement(result_score, 'textString')
+
+    version.text = 'V1.0'
+    message_id.text = str(uuid.uuid4())
+    sourced_id.text = result_sourcedid
+    language.text = 'en'
+    text_string.text = str(score)
+
+    return etree.tostring(envelope, xml_declaration=True, encoding='UTF-8')
+
+
+def sign_and_send_replace_result(assignment, xml):
+    """
+    Take the XML document generated in generate_replace_result_xml, and sign it
+    with the consumer key and secret assigned to the consumer. Send the signed
+    message to the LTI consumer.
+    """
+    outcome_service = assignment.outcome_service
+    assert outcome_service, 'assignment.outcome_service guaranteed to be non-null'
+    consumer_key = outcome_service.consumer_key
+    assert consumer_key, 'outcome_service.consumer_key guaranteed to be non-null'
+
+    # Fetch the secret associated with the consumer key
+    try:
+        consumer_secret = LtiConsumer.objects.get(
+            consumer_key=consumer_key
+        ).consumer_secret
+    except ObjectDoesNotExist:
+        log.error(
+            "Outcome Service: Can't retrieve consumer secret for key %s.",
+            consumer_key
+        )
+        return None
+
+    # Calculate the OAuth signature for the replace_result message.
+    # TODO: According to the LTI spec, there should be an additional
+    # oauth_body_hash field that contains a digest of the replace_result
+    # message. Testing with Canvas throws an error when this field is included.
+    # This code may need to be revisited once we test with other LMS platforms,
+    # and confirm whether there's a bug in Canvas.
+    oauth = requests_oauthlib.OAuth1(consumer_key, consumer_secret)
+
+    headers = {'content-type': 'application/xml'}
+    response = requests.post(
+        assignment.outcome_service.lis_outcome_service_url,
+        data=xml,
+        auth=oauth,
+        headers=headers
+    )
+    return response
+
+
+def check_replace_result_response(response):
+    """
+    Parse the response sent by the LTI consumer after an score update message
+    has been processed. Return True if the message was properly received, or
+    False if not. The format of this message is defined in the LTI 1.1 spec.
+    """
+    if response.status_code != 200:
+        log.error(
+            "Outcome service response: Unexpected status code %s",
+            response.status_code
+        )
+        return False
+
+    # etree can't handle XML that declares character encoding. Strip off the
+    # <?xml...?> tag and replace it with one that etree can parse
+    xml = response.text
+    if xml.lower().startswith('<?xml'):
+        xml = '<?xml version="1.0"?>' + xml.split('?>', 1)[1]
```

Comment:
Wait, really? When you tried it, were you feeding it bytes (as opposed to a unicode object)?